deps/thpool/thpool.c

```diff
@@ -1024,3 +1065,83 @@ static void redisearch_thpool_broadcast_new_state(redisearch_thpool_t *thpool,
   /* Wait on for the threads to pass the barrier and then destroy the barrier*/
   barrier_wait_and_destroy(&barrier);
 }
+
+
+/* ========================== CONFIGURATION THREAD REDUCTION ============================ */
+
+static void admin_job_change_state_last_destroys_barrier(void *job_arg_) {
+  adminJobArg *job_arg = job_arg_;
+  struct SignalThreadCtxWithBarrier *signal_struct = job_arg->arg;
+  ThreadState new_state = signal_struct->new_state;
+  job_arg->thread_ctx->thread_state = new_state;
+
+  /* Wait all threads to get the barrier */
+  int ret = pthread_barrier_wait(signal_struct->barrier);
+  if (ret == PTHREAD_BARRIER_SERIAL_THREAD) {
+    pthread_barrier_destroy(signal_struct->barrier);
+    rm_free(signal_struct->barrier);
+    rm_free(signal_struct);
+  }
+}
+
+static void config_reduce_threads_job(void *arg) {
+  struct config_reduce_threads_job_arg *job_arg = arg;
+  redisearch_thpool_t *thpool = job_arg->thpool;
+  size_t n_threads_to_remove = job_arg->n_threads_to_remove;
+  /* Create jobs and their args. */
+  redisearch_thpool_work_t jobs[n_threads_to_remove];
+  /* Create a barrier. */
+  pthread_barrier_t *barrier = rm_malloc(sizeof(pthread_barrier_t));
+  pthread_barrier_init(barrier, NULL, n_threads_to_remove);
+  struct SignalThreadCtxWithBarrier *signal_job_arg = rm_malloc(sizeof(struct SignalThreadCtxWithBarrier));
+  signal_job_arg->barrier = barrier;
+  signal_job_arg->new_state = THREAD_TERMINATE_ASAP;
```

Comment:
pass this as argument, as if workers is set to 0, it should terminate when Empty